### DISCUSSION TITLE
ESP32 DMX code

### DISCUSSION THREAD

**Discussion by SJKimpton** [someweisguy/esp_dmx#50] [General]
Hi Mitch,

I used your code (without understanding it much!) on a project about a year ago but find I can't update it because it now gives a compile error "dmx_config_t was not declared in this scope".  
I therefore just copied your DMXRead.ino example and find that it gives the same error on line 55.  I guess this is the result of some library update but I don't know how to proceed.  I'd be grateful for any help you can give.
Thanks
Steve

**Comment by someweisguy** [comment]
Thanks for the feedback!

Which version of esp_dmx you are using? The latest version, v3, is not compatible which v1 or v2. It sounds like you may be trying to use v3 example code with v1 or v2 code.

**Comment by SJKimpton** [comment]
Turns out I was using the most recent. I guess from just clicking 'yes' to IDE prompts that library updates were available.  
I worked out which library  version I would have been using when I wrote the code originally, regressed to that and everything now works correctly - Great!

I think your analysis may be back to front.  The example code from Github that verified the problem doesn't work with the V3 library but does with V1.  But thanks anyway.

**Comment by Florida-Rob** [comment]
Hi Mitch
I'm coming in cold here (to ESP32), sorry for the neub question
trying to switch micros, and considering the ESP32, but need RDM working perfectly...
I've looked at several ESP32 DMX-RDM code, and trying to figure out if the ESP32 can reliably be used for DMX-RDM with one of its own serial ports.
I need a full DMX-RDM solution, and see there are several RDM '[to-do](https://github.com/users/someweisguy/projects/5/views/1)' - says its 'in-progress'
How far are you off finishing DMX-RDM on the ESP32 ? #49, #16
and thank you in advance for your hard work here.  much appreciated.

Thanks, Rob

**Reply by someweisguy** [reply]
Thanks for the question! It certainly isn't a noob question at all. 

I, unfortunately, don't have a timeline for full RDM implementation. It depends on what you mean by "full DMX-RDM." Right now, this library is able to support the minimum requirements for RDM transmission. Therefore you could say that this library fully supports RDM transmission. However, there are over 50 different RDM messages that can be sent and most of them aren't yet supported by this library yet so you might say that this library does not fully support RDM transmission. Even with the minimum RDM requirements currently supported there are a few edge cases that this library does not support. 

I can say with confidence that the ESP32 can support RDM. I have done a fair amount of testing to ensure that the ESP32 can meet the timing requirements to send RDM responses and it looks possible!

I am sorry that I don't have a better answer, but hopefully this helps a bit!

**Comment by Florida-Rob** [comment]
I’m familiar with DMX for adding Parameters

What I really don’t want to do is to pull the code and start working on something that someone else is already fixing or is fixed in another branch ?

Is RDM slave / receive fully supported ?

What about sensors ?

And are Queued Messages supported ?

 

What about the issues raised #49 <https://github.com/someweisguy/esp_dmx/pull/49>  – RDM respond to discovery once per power/reset cycle?

What about #49 adding function rdm_uid_array_is_broadcast() to improve UID parsing ? was all ebeam-bob code now merged ? or do I need to fish through bobs’ code ?

 

How about #16 <https://github.com/someweisguy/esp_dmx/issues/16>  – are those changes included now ? (to fix the out and back to RDM ? i.e respond to all RDM messages – was the issue ‘muted’, is this fixed ? is bobs sp_rdm.c  integrated now ?

What about improvements from arneboe for usermod_rdm_dmx.h <https://github.com/arneboe/WLED/blob/updated/usermods/rdm_dmx/usermod_rdm_dmx.h>  or  WLED <https://github.com/netmindz/WLED/tree/DMX-Input-esp_dmx>  code ?

 

Is the latest code with all fixes all integrated in the Zip of v3.02 Beta ?

Is there an up-to-date to-do of fixes in flight and timeline ?

 

What about the USB driver to a PC for compatibility to OLA <https://www.openlighting.org/ola/get-help/ola-faq/>  and QLC+ ?  do we have to use a FTDI chip ? to behave as a ENTED DMX USB PRO

 

Rob

**Reply by someweisguy** [reply]
> Is RDM slave / receive fully supported ? What about sensors ? And are Queued Messages supported ?

No to all of those questions, unfortunately. I'm in the middle of my busy season so it has been difficult for me to find the motivation and time to work on these features. I believe that [DmxSerial2](https://github.com/mathertel/DmxSerial2) supports RDM supports sensors, but I am not sure about that off the top of my head. 

This library does currently support handling RDM timing requirements. Therefore it is possible for you to send/receive RDM if you hardcode the responses. I understand that isn't very helpful, but I'm afraid it's where I am at for now. Sorry for the inconvenience!

**Comment by SJKimpton** [comment]
Hi Mitch,
I instigated this discussion and you pointed out that I was using incompatible versions of the example code and the library.  I couldn't make the then most recent version of the example code work with the new library so I went back to V1.1.3 and everything was fine as far as I could see.  
In a new project I now have a problem [Guru meditation error Core 1 panic'd  (cache disabled but cached memory region accessed). ] which I think is caused by a combination of the (old) esp_dmx library and the esp32's Preferences library.  I'm sure the first thing to do is upgrade to the new library (3.0.2-beta ?) but when I copy the example code from Github there is no output.  
If I insert the line 'Serial.print(".")'  at the start of the loop, I  get a dot every second so I guess receive is not working.

Any advice?

**Reply by someweisguy** [reply]
Hm... This sounds suspiciously like #62. I just published v3.0.3-beta which contains a few fixes. I'm hoping that the issue described in #51 and #53 is the problem you are running into. If so, then v3.0.3-beta should fix the problem!

> In a new project I now have a problem [Guru meditation error Core 1 panic'd (cache disabled but cached memory region accessed). ]

This is a problem that I am also confident is fixed in v3.0.3-beta! The issue is that cache is being disabled but the ESP32 is attempting to read from or write to cache anyway. This can occur when reading or writing to flash or when using the non-volatile storage (NVS) library. I added a section in the readme about the solution [here](https://github.com/someweisguy/esp_dmx#using-flash-or-disabling-cache)!

Let me know if this was helpful or if you have any other questions!

**Comment by SJKimpton** [comment]
Hi Mitch,
 
First, an apology.  I’m new to Github and didn’t realise I should have raised an issue instead of a discussion.  
 
Second  - Thanks for your hard work.  3.0.3b works perfectly!
 
Regards
Steve

**Reply by someweisguy** [reply]
No worries! I'm glad to hear the new version is working out for you. Let me know if you have any more questions or concerns!
